libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,476 @@
+
+proc handshakeXXInbound(p: Noise, conn: Connection, p2pProof: ProtoBuffer): Future[tuple[cs1, cs2: CipherState; remotePrologue: seq[byte]; rs: Curve25519Key]] {.async.} =
+  const initiator = false
+
+  var hs: HandshakeState
+  init hs.ss
+
+  var prologue = p2pProof.buffer
+  hs.ss.mixHash(prologue)
+
+  hs.s.privateKey = p.noisePrivateKey
+  hs.s.publicKey = p.noisePublicKey
+
+  # -> e
+
+  var msg = await conn.readLp()
+
+  read_e()
+
+  let earlyData = hs.ss.decryptAndHash(msg)
+
+  # <- e, ee, s, es
+
+  msg.setLen(0)
+
+  write_e()
+  dh_ee()
+  write_s()
+  dh_es()
+
+  msg &= hs.ss.encryptAndHash(prologue)
+
+  await conn.writeLp(msg)
+
+  # -> s, se
+
+  msg = await conn.readLp()
+
+  read_s()
+  dh_se()
+
+  let remotePrologue = hs.ss.decryptAndHash(msg)
+
+  let (cs1, cs2) = hs.ss.split()
+  return (cs1, cs2, remotePrologue, hs.rs)
+
+proc handshake*(p: Noise, conn: Connection, initiator: bool): Future[NoiseConnection] {.async.} =
+  trace "Starting Noise handshake", initiator
+
+  # https://github.com/libp2p/specs/tree/master/noise#libp2p-data-in-handshake-messages
+  let
+    signedPayload = p.localPrivateKey.sign(PayloadString.getBytes & p.noisePublicKey.getBytes)
+    
+  var
+    libp2pProof = initProtoBuffer()
+    remoteProofBytes: seq[byte]
+    remoteKey: Curve25519Key
+    cs1: CipherState
+    cs2: CipherState
+
+  libp2pProof.write(initProtoField(1, p.localPublicKey))
+  libp2pProof.write(initProtoField(2, signedPayload))
+  # data field also there but not used!
+  libp2pProof.finish()
+
+  if initiator:
+    (cs1, cs2, remoteProofBytes, remoteKey) = await handshakeXXOutbound(p, conn, libp2pProof)
+  else:
+    (cs1, cs2, remoteProofBytes, remoteKey) = await handshakeXXInbound(p, conn, libp2pProof)
+
+  var
+    remoteProof = initProtoBuffer(remoteProofBytes)
+    remotePubKey: PublicKey
+    remoteSig: Signature
+  if remoteProof.getValue(1, remotePubKey) <= 0:
+    raise newException(NoiseHandshakeError, "Failed to deserialize remote public key.")
+  if remoteProof.getValue(2, remoteSig) <= 0:
+    raise newException(NoiseHandshakeError, "Failed to deserialize remote public key.")
+
+  let verifyPayload = PayloadString.getBytes & remoteKey.getBytes
+  if not remoteSig.verify(verifyPayload, remotePubKey):
+    raise newException(NoiseHandshakeError, "Noise handshake signature verify failed.")
+ 
+  if initiator:
+    let pid = PeerID.init(remotePubKey)
+    if not conn.peerInfo.peerId.validate():
+      raise newException(NoiseHandshakeError, "Failed to validate peerId.")
+    if pid != conn.peerInfo.peerId:
+      raise newException(NoiseHandshakeError, "Noise handshake, peer infos don't match! " & $pid & " != " & $conn.peerInfo.peerId)
+
+  var secure = new NoiseConnection
+  secure.insecure = conn
+  if initiator:
+    secure.readCs = cs2
+    secure.writeCs = cs1
+  else:
+    secure.readCs = cs1
+    secure.writeCs = cs2
+ 
+  unimplemented()
+
+method readExactly*(s: NoiseConnection,
```

Comment:
This doesn't need to be re-implemented here. For an example of how it should be plugged in, please see [`handleConn`]( https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/secure/secio.nim#L433-L448) and [`readLoop`](https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/secure/secio.nim#L417-L431).

What needs to happen is: 
- when a connection arrives, you should be able to handshake it and afterward read and write from the encrypted connection
- similarly when initiating the connections, the secure channel is negotiated and then the connection can be read and written to. 

`Secure` protocols never return the secured connection, instead they return a proxy `Connection` that wrapps a `BufferStream`. The `BufferStream` is a type that allows to push data (with `pushTo`) for it's readers, and register a handler to intercept writes from it's writers. 

If you take a look at the `readLoop` method linked above, you'll see that there is in fact a while loop that reads data from the secured connection using it's `readMessage` method, and then pushes the bytes to the proxied `BufferStream`, similarly in `handleConn` there is a [`writeHandler`](https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/secure/secio.nim#L435-L437) that calls secio's `writeMessage` method to write to the underlying secured connection.

This is essentially what needs to happen in order to plug you implementation into libp2p. 

I'm thinking now, that we can actually move the logic I just described into the base `Secure` type and define some common interface such as `handshake`, `readMessage` and `writeMessage` that secure protocols expose so we don't have to reimplement this for each secure protocol.

Reply:
Hmm I see, I could not see such proxy implementation in other libp2p implementations, at least not at protocol level, probably moving it to `Secure` is indeed what we should do

Reply:
Yeah, there is little overlap between implementations given the vastly different primitives each language offers, tho you can find some resemblance with the js implementation given that both use `async` to some extent.

Reply:
On the integration front, what we need to do is `handleConn` and `readLoop` to `secure.nim` and then add a base `SecureConnection` type that implement `readMessage` and `writeMessage`. All of this is already implemented, in secio it's just a matter of moving things around.

Reply:
I added a generic readLoop that consumes read/write message (those don't need to be methods, since generic will resolve them at compile time)